capa/features/extractors/dnfile/helpers.py

```diff
@@ -299,22 +327,79 @@ def get_dotnet_unmanaged_imports(pe: dnfile.dnPE) -> Iterator[DnUnmanagedMethod]
         # like kernel32.CreateFileA
         yield DnUnmanagedMethod(token, module, method)
 
-
+def get_nested_class_table(pe):
+    nested_class_table = {}
+    
+    # Used to find nested classes in typedef
+    for rid, nestedclass in iter_dotnet_table(pe, dnfile.mdtable.NestedClass.number):
+        assert isinstance(nestedclass, dnfile.mdtable.NestedClassRow)
+        nested_class_table[nestedclass.NestedClass.row_index] = nestedclass.EnclosingClass.row_index
+
+    return nested_class_table
+
+def typedef_helper(index, nested_class_table, typedef_class_table, typedef_name, name):
+    # Append the current typeref name
+    typedef_name.append(name)
+    
+    while nested_class_table[index] in nested_class_table:
+        name = typedef_class_table[nested_class_table[index]-1].TypeName
+        typedef_name.append(name)
+        index = nested_class_table[index]
+        
+    # Document the root enclosing details
+    enclosing_name = typedef_class_table[nested_class_table[index]-1].TypeName
+    typedef_name.append(enclosing_name)
+    namespace = typedef_class_table[nested_class_table[index]-1].TypeNamespace
+    
+    return namespace, tuple(typedef_name[::-1])
+
+def typeref_helper(index, typeref_table, typeref_name, name):
+    # Not appending the current typeref name to avoid potential duplicate
+    
+    while type(typeref_table[index - 1].ResolutionScope.table) is dnfile.mdtable.TypeRef:
+        # Recursively call helper function with enclosing typeref details
+        typeref_name.append(name)
+        name = typeref_table[index - 1].TypeName
```

Comment:
The `get_dotnet_table_row` function should be substituted everywhere that you directly index a table by row index. This will require adding additional checks to ensure the object returned is not `None` before using it.

Reply:
Just pushed an implementation that works locally. Will need to harden it with some test cases. The implementation borrows logic from other functions and should properly account for when the object is `None`.

Reply:
Patched the implementation to remove an extra check and correct index value handling. All tests on dotnet data set pass.